src/elf.c

```diff
@@ -8,245 +8,237 @@
 /* ELF file manipulation */
 
 int elf_symbol_index;
-int elf_symtab_index;
-int elf_strtab_index;
-int elf_section_index;
 
-void elf_write_section_str(char *vals, int len)
+void elf_write_str(strbuf_t *elf_array, char *vals)
 {
-    for (int i = 0; i < len; i++)
-        elf_section[elf_section_index++] = vals[i];
+    /*
+     * Note that strbuf_puts() does not push the null character.
+     *
+     * If necessary, use elf_write_byte() to append the null character
+     * after calling elf_write_str().
+     */
+    strbuf_puts(elf_array, vals);
 }
 
-void elf_write_data_str(char *vals, int len)
+void elf_write_byte(strbuf_t *elf_array, int val)
 {
-    for (int i = 0; i < len; i++)
-        elf_data[elf_data_idx++] = vals[i];
-}
-
-void elf_write_header_byte(int val)
-{
-    elf_header[elf_header_idx++] = val;
-}
-
-void elf_write_section_byte(char val)
-{
-    elf_section[elf_section_index++] = val;
+    strbuf_putc(elf_array, val);
 }
 
 char e_extract_byte(int v, int b)
 {
-    return (v >> (b * 8)) & 0xFF;
+    return (v >> (b << 3)) & 0xFF;
 }
 
-int elf_write_int(char *buf, int index, int val)
+void elf_write_int(strbuf_t *elf_array, int val)
```

Comment:
How about changing the function prototype to `strbuf_t *elf_write_int(int val)`? That is, we can check its return value for initialization.